proposals/BVAL-508.md

```diff
@@ -0,0 +1,257 @@
+---
+title: Validating elements contained in a container (like collections)
+layout: default
+author: Emmanuel Bernard
+comments: true
+---
+
+# #{page.title}
+
+[Link to JIRA ticket][jira]  
+[Related JIRA](https://hibernate.atlassian.net/browse/BVAL-499)  
+
+## Problem
+
+It is useful to be able to apply constrains on elements contained in a so called _container_.
+Examples of containers are:
+
+* `Collection`
+* `Optional`
+* JavaFX's `Property`
+
+But today the constraints are applied on the container value itself and not its content.
+
+## Proposition
+
+### Containers
+
+Bean Validation will offer the notion of container of values to validate.
+Containers can contain one value, or several values.
+Extractors for specific containers can be implemented.
+
+TODO: should we differentiate single and plural?
+
+    // Draft proposal for the contract
+    public interface SingleContainerValueExtractor<CONTAINER,CONTAINED> {
+        CONTAINED extractValue(CONTAINER container);
+    }
+    
+    public interface ManyContainerValuesExtractor<CONTAINER,CONTAINED> {
+        Iterable<CONTAINED> extractValue(CONTAINER container);
+    }
+    
+    // Examples
+    
+    //TODO can T remain unspecified? or should it be ? or Object?
+    public class OptionalContainerExtractor<T> implements SingleContainerValueExtractor<Optional<T>,T> {
+        T extractValue(Optional<T> optional) { ... };
+    }
+
+Bean Validation will have out of the box support for containers for `Iterable` and `Map`.
+
+TODO: any other Java SE type to consider?
+TODO: should we support JavaFX Property<T>?
+
+Custom container extractors can be provided via:
+
+* property
+* factory
+* service locator
+
+TODO: refine how the container implementations are discovered.
+See [Hibernate Validator's feature](http://docs.jboss.org/hibernate/validator/5.2/reference/en-US/html_single/#section-value-handling) 
+
+#### When are containers used?
+
+When a property is of a known container, constraints declarations are explored.
+A typical example is a property using constraints on type parameters of `TYPE_USE` (`Collection<@Size String>`).
+
+If constraints are discovered, the unwrapping mechanism is used
+to read contained value(s) and potentially apply constraints on them (see below).
+
+By constraints, we also mean property traversal `@Valid`.
```

Comment:
That's interesting.
In this proposal, the first would mean that each `PlushGiraffe` is validated wheas in the second, the collection object itself is explored for properties to validate.

But that's not how BV 1.1 describes things as collections are a special case.

I guess we need to describe it specifically. so that both forms are equivalent.
Agreed?